clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -149,6 +157,74 @@ public Sender(LogContext logContext,
         this.retryBackoffMs = retryBackoffMs;
         this.apiVersions = apiVersions;
         this.transactionManager = transactionManager;
+        this.inFlightBatches = new HashMap<>();
+    }
+
+    public Map<TopicPartition, List<ProducerBatch>> getInflightBatches() {
+        return inFlightBatches;
+    }
+
+    public List<ProducerBatch> inFlightBatches(TopicPartition tp) {
+        return inFlightBatches.containsKey(tp) ? inFlightBatches.get(tp) : new ArrayList<>();
+    }
+
+    public void maybeRemoveFromInflightBatches(ProducerBatch batch) {
+        List<ProducerBatch> batches = inFlightBatches.get(batch.topicPartition);
+        if (batches != null) {
+            batches.remove(batch);
+        }
+    }
+
+    /**
+     *  Get the in-flight batches that has reached delivery timeout.
+     */
+    private List<ProducerBatch> getExpiredInflightBatches(long now) {
+        Set<TopicPartition>  topicPartitions = this.accumulator.getTopicPartitions();
```

Comment:
Can you clarify why use this for the iteration? I would have expected we would iterate over the entry set of `inFlightBatches`.

Reply:
this is really not needed.  updated to iterate through `inFlightBatches` directly.